tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15169,6 +15169,60 @@ static bool nodeInputIsCompatible(DXIL::NodeIOKind IOType,
   }
 }
 
+// Diagnose input node record to make sure it has exactly one SV_DispatchGrid
+// semantics. Recursivelly walk all fields on the record and all of its base
+// classes/structs
+void DiagnoseMustHaveOneDispatchGridSemantics(Sema &S,
+                                              CXXRecordDecl *InputRecordDecl,
+                                              SourceLocation &DispatchGridLoc,
+                                              bool &Found) {
+  // Iterate over fields of the input record struct
+  for (auto FieldDecl : InputRecordDecl->fields()) {
+    // Check if any of the fields have SV_DispatchGrid annotation
+    for (const hlsl::UnusualAnnotation *it :
+         FieldDecl->getUnusualAnnotations()) {
+      if (it->getKind() == hlsl::UnusualAnnotation::UA_SemanticDecl) {
+        const hlsl::SemanticDecl *sd = cast<hlsl::SemanticDecl>(it);
+        if (sd->SemanticName.equals("SV_DispatchGrid")) {
+          if (!Found) {
+            Found = true;
+            DispatchGridLoc = it->Loc;
+          } else {
+            // There should be just one SV_DispatchGrid in per record struct
+            S.Diags.Report(
+                it->Loc,
+                diag::err_hlsl_dispatchgrid_semantic_already_specified);
+            S.Diags.Report(DispatchGridLoc, diag::note_defined_here)
+                << "other SV_DispatchGrid";
+          }
+          break;
+        }
+      }
+    }
+  }
+
+  // Walk up the inheritance chain and check all fields on base classes
+  for (CXXRecordDecl::base_class_iterator B = InputRecordDecl->bases_begin(),
+                                          BEnd = InputRecordDecl->bases_end();
+       B != BEnd; ++B) {
+
+    QualType BaseType = B->getType();
+    CXXRecordDecl *BaseTypeDecl =
```

Comment:
I'm worried about if this dyn_cast assigns a nullptr to BaseTypeDecl, or if any caller that calls this  recursive function passes a nullptr. This is because the for loop at the top of the function would immediately dereference the pointer. Would it be possible to just do a nullptr check at the top and return early if BaseTypeDecl is null?